numpy/ma/core.py

```diff
@@ -3873,77 +3875,87 @@ def _delegate_binop(self, other):
-                check = check.view(type(self))
-                check._mask = self._mask
-            except AttributeError:
-                # Dang, we have a bool instead of an array: return the bool
-                return check
+        smask = self._mask
+        if smask is nomask:
+            mask = omask
         else:
-            odata = filled(other, 0)
-            check = self.filled(0).__eq__(odata).view(type(self))
-            if self._mask is nomask:
-                check._mask = omask
-            else:
-                mask = mask_or(self._mask, omask)
-                if mask.dtype.names:
-                    if mask.size > 1:
-                        axis = 1
-                    else:
-                        axis = None
-                    try:
-                        mask = mask.view((bool_, len(self.dtype))).all(axis)
-                    except ValueError:
-                        mask = np.all([[f[n].all() for n in mask.dtype.names]
-                                       for f in mask], axis=axis)
-                check._mask = mask
+            mask = mask_or(smask, omask)
+
+        odata = other.data if isinstance(other, MaskedArray) else other
+        if mask.dtype.names and mask is not nomask:
+            # For possibly masked structured arrays we need to be careful,
+            # since the standard structured array comparison will use all
+            # elements, masked or not.  To avoid this, we set all masked
+            # elements in self to other, so they'll count as equal.
+            # To prepare, we ensure we have the right shape.
+            sbroadcast = np.broadcast_to(self, np.broadcast(self, other).shape,
+                                         subok=True)
+            sbroadcast._mask = mask
+            sdata = sbroadcast.filled(odata)
+            # Now take care of the mask; the merged mask should have an item
+            # masked if all elements were masked (in one and/or other).
+            all_masked = np.ones(1, mask.dtype)
+            mask = (mask == all_masked)
+
+        else:
+            # For regular arrays, just use the data as they come.
+            sdata = self.data
+
+        check = compare(sdata, odata)
+
+        if isinstance(check, (bool_, bool)):
+            return masked if mask else check
+
+        if mask is not nomask:
+            # Adjust elements that were masked, which should be treated
+            # as equal if masked in both, unequal if masked in one.
+            # Note that this works automatically for structured arrays too.
+            check[mask] = compare(smask, omask)[mask]
+
+        check = check.view(MaskedArray)
+        check._mask = mask
         return check
 
-    def __ne__(self, other):
+    def __eq__(self, other):
+        """Check whether other equals self elementwise.
+
+        When either of the elements is masked, the result is masked as well,
+        but it is still set, with self and other considered equal if both
```

Comment:
I'm not sure what you mean by "it is still set"?

What about just something like "the resulting element is masked if either input element is masked" for this part?